### PR TITLE
Optimize sequence reordering on TPU by eliminating jnp.take gather

### PR DESCRIPTION
**Pull request by zcjhao**
# Description
  
  Replaces the general `jnp.take` (gather) operation in `reorder_sequence` with structured strided slicing and JAX layout operations. 
  
  **Context & Problem:** 
  During training sweeps of DeepSeek-V3.2 at long contexts (128K) with CP=128, the sequence reordering logic was identified as a performance bottleneck. XLA was lowering the original `jnp.take` generalized gather into expensive sequential TensorCore `while` loops. 

  **Solution:**
  Because the permutation indices are perfectly deterministic based on `cp_size`, we can natively recreate the permutation mapping using contiguous layout transformations (for load balancing folding) and stride-2 vector slicing (for restoring original order). 

  # Tests

  Testing performed locally via `pytest`:
  
  1. **Mathematical Equivalence Sweep:** Added and ran`test_reorder_equivalence_sweep` in `tests/unit/max_utils_test.py`. It sweeps `cp_size`, direction (`to_contiguous=True/False`), tensor ranks (1D, 4D), and negative dimension indexing (`seq_dim = -1`, `seq_dim = -3`), asserting bit-for-bit mathematical equivalence against the original gather-based implementation.
  2. **Lossless Roundtrip Verification:** Added`test_reorder_roundtrip` in `tests/unit/max_utils_test.py` to ensure that load-balancing a tensor and subsequently restoring it yields the exact original input tensor.
  
# Profiles
- [Baseline](https://xprof.corp.google.com/overview_page/zjiahao-13925005173721863237): Step time 77s
- [Improved xprof](https://xprof.corp.google.com/overview_page/zjiahao-16358326198314446624): Step time 74s

  # Checklist

  Before submitting this PR, please make sure (put X in square brackets):
  - [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
  - [x] I have necessary comments in my code, particularly in hard-to-understand areas.
  - [x] I have run end-to-end tests tests and provided workload links above if applicable. *(Update this to [x] once cluster validation
finishes!)*
  - [x] I have made or will make corresponding changes to the doc if needed.

  TAG=agy
  CONV=f2dcd166-f4f6-47ab-8984-0f5cdfca2ce1

### CODE REVIEW

File: src/maxtext/utils/max_utils.py [hunk 1]
```diff
@@ -871,46 +859,30 @@ def reorder_sequence(tensor, cp_size: int, seq_dim: int = 1, to_contiguous: bool
...
 
-  if not to_contiguous:
-    # Create first and second halves
-    first_half = jnp.arange(cp_size)
-    second_half = jnp.arange(2 * cp_size - 1, cp_size - 1, -1)
-
-    # Stack and reshape to interleave
-    src_indices = jnp.stack([first_half, second_half], axis=1).reshape(-1)
+  # Swap seq_dim with axis 0 to perform slicing/concat easily
+  swapped = jnp.swapaxes(reshaped, 0, seq_dim)
```
Comment (gobbleturk) [inline src/maxtext/utils/max_utils.py:890]: add a comment for shape  - e.g,.is it b, t, d?

**Review by gobbleturk — APPROVED** [review]
Awesome!

**Review by huytransformer — APPROVED** [review]
Excellent work!!!

File: src/maxtext/utils/max_utils.py (hunk 1 above)
Comment (RissyRan) [reply src/maxtext/utils/max_utils.py:890]: +1, let's add a comment on shape of this swapped variable for readability.

File: src/maxtext/utils/max_utils.py [hunk 2]
```diff
@@ -859,46 +859,32 @@ def reorder_sequence(tensor, cp_size: int, seq_dim: int = 1, to_contiguous: bool
   if seq_len % (cp_size * 2) != 0:
     raise ValueError(f"{tensor.shape=} is not a multiple of {cp_size*2=}")
 
-  # [B, S, H, D]: [B, 2*cp_size, S/2*cp_size, H, D] -> [B, 2, S/2*cp_size, H, D]
-  # [S, B, H, D]: [2*cp_size, S/2*cp_size, B, H, D] -> [2, S/2*cp_size, B, H, D]
   ori_tensor_shape = tensor.shape
+
   reshaped = tensor.reshape(
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/maxtext/utils/max_utils.py:881]: Could you consider this comment and add a test? i.e. it's possible to use Axis -1 refers to the very last dimension

**Review by RissyRan — CHANGES_REQUESTED** [review]
Thanks for the change! A few minor comments.

**Comment by zcjhao** [comment]
@RissyRan Applied the changes. Lmk if it looks good to you!

**Review by RissyRan — APPROVED** [review]
